File: ShowPaiGowLayer.js
var FRAME_RATE = 60;
var BOUNCE_TIME = 0.2;
var	BounceDirectionStayingStill = 0;
var	BounceDirectionGoingLeft = 1;
var	BounceDirectionGoingRight = 2;

var ShowPaiGowLayer = cc.Layer.extend({
	isDragging: false,
	lasty: 0,
	xvel: 0,
	direction: BounceDirectionStayingStill,
	scrollLayer: null,
    ctor:function () {
    },
    init:function (rowNum) {
        var bRet = false;
        if (this._super()) {
			this.setTouchEnabled(true);
            this.isDragging = false;
            this.lasty = 0.0;
            this.xvel = 0.0;
            this.direction = BounceDirectionStayingStill;
            var paiGowArray = CFSModePaiGowManager.getInstance().paiGowObjectArrayForRow(rowNum);
            var tempPaiGowObject = paiGowArray[0];
            this.scrollLayer = cc.LayerColor.create(cc.c4b(22, 144, 227, 200), (paiGowArray.length + 1 ) * 38, PaiGowObject.heightForPaiGow());
            this.scrollLayer.setPosition(cc.p(0, tempPaiGowObject.pg_position.y - PaiGowObject.heightForPaiGow() / 2));
            this.addChild(this.scrollLayer, 0);

            for (var i = 0; i < paiGowArray.length; i++) {
                var paiGowObject = paiGowArray[i];
                var copySprite = new cc.Sprite();
                copySprite.initWithSpriteFrameName(paiGowObject.pg_name);
                copySprite.setPosition(cc.p(paiGowObject.pg_colNum * 38, PaiGowObject.heightForPaiGow() / 2));
                this.scrollLayer.addChild(copySprite, 2);
            }
			var closeSprite = cc.Sprite.create();
			closeSprite.initWithSpriteFrameName("close.png");
            closeSprite.setScale(0.7);
			var menuItem = cc.MenuItemSprite.create(closeSprite, null, null, this.closeLayer, this);
			var menu = cc.Menu.create(menuItem, null);
			menu.setPosition(cc.p(14, 60));
            this.scrollLayer.addChild(menu, 2);
			
			this.schedule(this.onTimerUpdate);
            bRet = true;
        }
        return bRet;
    },
    closeLayer: function (sender) {
    	this.removeFromParent(true);
    },
    onTimerUpdate: function (dt) {
		var pos = this.scrollLayer.getPosition();
		var right = pos.x + this.getBoundingBox().origin.x + this.scrollLayer.getContentSize().width;
		var left = pos.x + this.getBoundingBox().origin.x;
		var minX = this.getBoundingBox().origin.x;
		var maxX = this.getBoundingBox().origin.x + this.getBoundingBox().size.width;
		if(!this.isDragging) {
			var friction = 0.96;
			if(left > minX && this.direction != BounceDirectionGoingLeft) {
                this.xvel = 0;
                this.direction = BounceDirectionGoingLeft;
			}
			else if(right < maxX && this.direction != BounceDirectionGoingRight)	{
                this.xvel = 0;
                this.direction = BounceDirectionGoingRight;
			}
			if(this.direction == BounceDirectionGoingRight) {
				if(this.xvel >= 0) {
					var delta = (maxX - right);
                    this.xvel = (delta / (BOUNCE_TIME * FRAME_RATE));
				}
				if((right + 0.5) == maxX) {
					pos.x = right -  this.scrollLayer.getContentSize().width;
                    this.xvel = 0;
                    this.direction = BounceDirectionStayingStill;
				}
			}
			else if(this.direction == BounceDirectionGoingLeft) {
				if(this.xvel <= 0) {
					delta = (minX - left);
                    this.xvel = (delta / (BOUNCE_TIME * FRAME_RATE));
				}
				if((left + 0.5) == minX) {
					pos.x = left - this.getBoundingBox().origin.x;
                    this.xvel = 0;
                    this.direction = BounceDirectionStayingStill;
				}
			}
			else {
                this.xvel *= friction;
			}
			pos.x += this.xvel;
            this.scrollLayer.setPosition(pos);
		}
		else {
			if(left <= minX || right >= maxX) {
                this.direction = BounceDirectionStayingStill;
			} 
			if(this.direction == BounceDirectionStayingStill) {
                this.xvel = (pos.x - this.lasty)/2;
                this.lasty = pos.x;
			}
		}
    },
    onTouchesBegan:function (touches, event) {
    	var location = touches[0].getLocation();
		if (!cc.Rect.CCRectContainsPoint(this.scrollLayer.getBoundingBox(), location)) {
			return true;
		}
        this.isDragging = true;
		return true;
    },
    onTouchesMoved:function (touches, event) {
    	var location = touches[0].getLocation();
		if (!cc.Rect.CCRectContainsPoint(this.scrollLayer.getBoundingBox(), location)) {
			return true;
		}
		var preLocation = touches[0].getPreviousLocation();
        var curLocation = location;
        var nowPosition = this.scrollLayer.getPosition();
		nowPosition.x += ( curLocation.x - preLocation.x );
        this.scrollLayer.setPosition(nowPosition);
    },
    onTouchesEnded:function (touches, event) {
        this.isDragging = false;
    }
});

ShowPaiGowLayer.create = function (rowNum) {
    var sg = new ShowPaiGowLayer();
    if (sg && sg.init(rowNum)) {
        return sg;
    }
    return null;
};

File: ClearFiveStages-iOS/ClearFiveStages/Scene/ShowPaiGowLayer.m
#import "ShowPaiGowLayer.h"
#import "CFSModePaiGowManager.h"

@implementation ShowPaiGowLayer

+(id) layerWithRow:(int)rowNum {
    return [[[self alloc] initWithRow:rowNum] autorelease];
}

-(id) initWithRow:(int)rowNum {
	if ((self = [super initWithColor:ccc4(255, 255, 255, 0)])) {
		[self setTouchEnabled:YES];
		isDragging = NO;
		lasty = 0.0f;
		xvel = 0.0f;
		direction = BounceDirectionStayingStill;
		scrollLayer = [CCLayerColor layerWithColor:ccc4(22, 144, 227, 200)];
		
		NSMutableArray *paiGowArray = [NSMutableArray arrayWithArray:[[CFSModePaiGowManager sharedManager] paiGowRowArrayForRow:rowNum]];
        PaiGowObject *paiGowObject = paiGowArray[0];
		scrollLayer.contentSize = CGSizeMake((paiGowArray.count + 1 ) * 38, [PaiGowObject heightForPaiGow]);
		scrollLayer.position = ccp(0, paiGowObject.pg_position.y - [PaiGowObject heightForPaiGow] / 2);
		[self addChild: scrollLayer z:0];
        for (int i = 0; i < paiGowArray.count; i++) {
			PaiGowObject *paiGowObject = paiGowArray[i];			
			CCSprite *copySprite = [CCSprite spriteWithSpriteFrameName:paiGowObject.pg_name];
			copySprite.position = ccp(paiGowObject.pg_colNum * 38, [PaiGowObject heightForPaiGow] / 2);
			[scrollLayer addChild:copySprite z:2];
		}
		CCSprite *close = [CCSprite spriteWithSpriteFrameName:@"close.png"];
		close.scale = 0.7f;
		CCMenuItemSprite *menuItem = [CCMenuItemSprite itemWithNormalSprite:close selectedSprite:nil target:self selector:@selector(closeLayer)];
		CCMenu *menu = [CCMenu menuWithItems:menuItem, nil];
		menu.position = ccp(14, 60);
		[scrollLayer addChild:menu z:2];
		[self scheduleUpdate];
	}
	return self;
}

-(void) closeLayer {
	[self removeFromParentAndCleanup:YES];
}

-(void) update:(ccTime)delta {
	CGPoint pos = scrollLayer.position;	
	float right = pos.x + [self boundingBox].origin.x + scrollLayer.contentSize.width;
	float left = pos.x + [self boundingBox].origin.x;
	float minX = [self boundingBox].origin.x;
	float maxX = [self boundingBox].origin.x + [self boundingBox].size.width;	
	if(!isDragging) {
		static float friction = 0.96f;
		if(left > minX && direction != BounceDirectionGoingLeft) {
			xvel = 0;
			direction = BounceDirectionGoingLeft;
		}
		else if(right < maxX && direction != BounceDirectionGoingRight)	{
			xvel = 0;
			direction = BounceDirectionGoingRight;
		}
		if(direction == BounceDirectionGoingRight) {
			if(xvel >= 0) {
				float delta = (maxX - right);
				float yDeltaPerFrame = (delta / (BOUNCE_TIME * FRAME_RATE));
				xvel = yDeltaPerFrame;
			}
			if((right + 0.5f) == maxX) {
				pos.x = right -  scrollLayer.contentSize.width;
				xvel = 0;
				direction = BounceDirectionStayingStill;
			}
		}
		else if(direction == BounceDirectionGoingLeft) {			
			if(xvel <= 0) {
				float delta = (minX - left);
				float yDeltaPerFrame = (delta / (BOUNCE_TIME * FRAME_RATE));
				xvel = yDeltaPerFrame;
			}
			if((left + 0.5f) == minX) {
				pos.x = left - [self boundingBox].origin.x;
				xvel = 0;
				direction = BounceDirectionStayingStill;
			}
		}
		else {
			xvel *= friction;
		}
		pos.x += xvel;	
		scrollLayer.position = pos;
	}
	else {
		if(left <= minX || right >= maxX) {	
			direction = BounceDirectionStayingStill;
		}
		if(direction == BounceDirectionStayingStill) {
			xvel = (pos.x - lasty)/2;
			lasty = pos.x;
		}
	}
}

-(void) registerWithTouchDispatcher {
	[[CCDirector sharedDirector].touchDispatcher addTargetedDelegate:self priority:0 swallowsTouches:NO];
}

-(BOOL) ccTouchBegan:(UITouch*)touch withEvent:(UIEvent *)event {
	CGPoint location =  [[CCDirector sharedDirector] convertToGL:[touch locationInView: [touch view]]];
	if ( !CGRectContainsPoint([scrollLayer boundingBox], location)) return YES;
	isDragging = YES;	
	return YES;
}

-(void) ccTouchMoved:(UITouch*)touch withEvent:(UIEvent *)event {
	CGPoint touchLocation = [touch locationInView: [touch view]];
	CGPoint location =  [[CCDirector sharedDirector] convertToGL:touchLocation];
	if ( !CGRectContainsPoint([scrollLayer boundingBox], location)) return;
	CGPoint a = [[CCDirector sharedDirector] convertToGL:[touch previousLocationInView:[touch view]]];
	CGPoint b = [[CCDirector sharedDirector] convertToGL:[touch locationInView:[touch view]]];
	
	CGPoint nowPosition = scrollLayer.position;
	nowPosition.x += ( b.x - a.x );
	scrollLayer.position = nowPosition;
}

-(void) ccTouchEnded:(UITouch*)touch withEvent:(UIEvent *)event {
	isDragging = NO;
}
@end